scripts/justmat/foulplay.lua

```diff
@@ -388,3 +477,131 @@ midi.add = function(dev)
   dev.event = clk.process_midi
   print("foulplay: midi device added", dev.id, dev.name)
 end
+
+-- grid stuff - junklight
+
+-- grid key function
+function gridkey(x, y, state)
+  -- print("x:",x," y:",y," state:",state)
+  -- use first column to switch track edit
+  if x == 1 then
+    track_edit = y
+    reer(track_edit)
+  end
+  -- second column provides mutes 
+  -- key presses just toggle
+  -- switches on grid up - so you can hold down
+  -- and lift to get timing right 
+  if x == 2 and state == 1 then
+    mutes[y] = not mutes[y]
+  end
+  -- 4-6,8 are used to open track parameters pages
+  if y == 8 and x >= 4 and x <= 7 and state == 1 then
+    view = 1
+    page = x - 4
+  else 
+    view = 0
+  end
+  -- copy button 
+  if x == 8 and y==8 and state == 1 then 
+    copy_mode = true
+    copy_source_x = -1
+    copy_source_y = -1
+  elseif x == 8 and y==8 and state == 0 then
+    copy_mode = false
+    copy_source_x = -1
+    copy_source_y = -1
+  end  
+  -- memory cells 
+  -- if we aren't in copy_mode 
+  -- press to switch memory 
+  -- switches on grid up not grid down
+  if not copy_mode then
+    if y >= 1 and y <= 5 and x >= 4 and x <= 8 and state == 1 then 
+      current_mem_cell = cellfromgrid(x,y)
+      current_mem_cell_x = x
+      current_mem_cell_y = y
+    end
+  else
+    if y >= 1 and y <= 5 and x >= 4 and x <= 8 and state == 0 then 
+      -- copy functionality 
+      -- if we are in copy mode then don't switch 
+      -- memories - copy about instead 
+      if copy_source_x == -1 then 
+        -- first button sets the source
+        copy_source_x = x
+        copy_source_y = y
+      else 
+        -- copy source into target 
+        if copy_source_x ~= -1 and not ( copy_source_x == x and copy_source_y == y) then
+          sourcecell = cellfromgrid( copy_source_x , copy_source_y )
+          targetcell = cellfromgrid( x , y )
+          memory_cell[targetcell] = simplecopy(memory_cell[sourcecell])
+        end
+      end
+    end
+  end
+  
+  redraw()
+end
+
+local function grid_redraw()
```

Comment:
i'm not sure but i would expect for this to be local, it would have to get defined before it's referenced in ` metro_grid_redraw`.   did it run as is?

Reply:
It does, but that is a quick change, so I will fix it.